### PR TITLE
Add test and basic implementation for formatter preview mode

### LINKED ISSUE #7440: Formatter incompatibility: ellipses are not inlined in `.py` files
Given:

```python
def f(): ...

def f():
    ...

def f():
    """Docstring"""
    ...
```

Recent versions of Black format as:

```python
def f(): ...


def f(): ...


def f():
    """Docstring"""
    ...
```

While Ruff formats as:

```python
def f():
    ...


def f():
    ...


def f():
    """Docstring"""
    ...
```

This changed in Black 23.9.0 (https://github.com/psf/black/pull/3796).

### PR DESCRIPTION
**Pull request by konstin**
**Summary** Prepare for the black preview style becoming the black stable style at the end of the year.

This adds a new test file to compare stable and preview on some relevant preview options in black, and makes `format_dev` understand the black preview flag. I've added poetry as a project that uses preview.

I've implemented one specific deviation (collapsing of stub implementation in non-stub files) which showed up in poetry for testing. This also improves poetry compatibility from 0.99891 to 0.99919.

Fixes #7440

New compatibility stats:
| project        | similarity index  | total files       | changed files     |
|----------------|------------------:|------------------:|------------------:|
| cpython        |           0.75803 |              1799 |              1647 |
| django         |           0.99983 |              2772 |                35 |
| home-assistant |           0.99953 |             10596 |               189 |
| poetry         |           0.99919 |               317 |                12 |
| transformers   |           0.99963 |              2657 |               332 |
| twine          |           1.00000 |                33 |                 0 |
| typeshed       |           0.99978 |              3669 |                20 |
| warehouse      |           0.99969 |               654 |                15 |
| zulip          |           0.99970 |              1459 |                22 |

### CODE REVIEW

File: crates/ruff_python_formatter/tests/snapshots/format@preview.py.snap [hunk 1]
```diff
@@ -1,13 +1,45 @@
 ---
 source: crates/ruff_python_formatter/tests/fixtures.rs
-input_file: crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/assign_breaking.py
+input_file: crates/ruff_python_formatter/resources/test/fixtures/ruff/preview.py
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/tests/snapshots/format@preview.py.snap:3]: Should we instead run every test under preview and non-preview?

**Comment by konstin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #8049** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/8049" target="_blank">Graphite</a> 
    * **PR #8044** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/8044" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/8044?utm_source=stack-comment).

File: crates/ruff_python_formatter/tests/snapshots/format@preview.py.snap (hunk 1 above)
Comment (konstin) [reply crates/ruff_python_formatter/tests/snapshots/format@preview.py.snap:3]: `hyperfine "cargo test -p ruff_python_formatter"` goes from 500ms to 580ms on my machine, acceptable for me. 

Having both versions makes it hard to see what actually changed given our large snapshots, so i opted for instead adding the diff as separate code block if there is one. I also changed all non-function/class tests to use `pass` instead as real code generally does, which i split out to #8049 to keep the diff readable.

Comment (konstin) [reply crates/ruff_python_formatter/tests/snapshots/format@preview.py.snap:3]: I would like to keep `preview.py` as-is though, it's a useful tool for tracking progress and comparing to black (i can diff both snapshots with what the black playground does)

File: crates/ruff_python_formatter/tests/fixtures.rs [hunk 2]
```diff
@@ -142,16 +142,40 @@ fn format() {
...
+            ensure_stability_when_formatting_twice(formatted, options.clone(), input_path);
 
-            writeln!(
-                snapshot,
-                "## Output\n{}",
-                CodeFrame::new("py", &formatted_code)
-            )
-            .unwrap();
+            // We want to capture the differences in the preview style in our fixtures
+            let options_preview = options.with_preview(PreviewMode::Enabled);
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/tests/fixtures.rs:150]: Should we skip this if `options.preview == Enabled` when using a test specific options file?

File: crates/ruff_python_formatter/tests/snapshots/format@statement__function.py.snap [hunk 3]
```diff
@@ -996,4 +996,167 @@ def default_arg_comments2(  #
 ```
 
 
+## Preview changes
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/tests/snapshots/format@statement__function.py.snap:999]: I wonder how noisy this gets once we start having many preview style rules. But let's see how it goes.

**Review by MichaReiser — APPROVED** [review]

**Comment by charliermarsh** [comment]
@konstin - Is this good to merge?

**Comment by konstin** [comment]
yep

**Pull request merged**